dali/sasha/saarch.cpp

```diff
@@ -83,297 +83,446 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
+            return true;
+        }
+        bool addOutputFromWUTree(StringArray &outputFields, IPropertyTree *wuTree)
+        {
+            if (outputXML)
+            { //cmd->getAction()==SCA_GET
+                StringBuffer xml;
+                toXML(wuTree, xml);
+                if (!cmd->addResult(xml.str()))
+                    return false; //Log overflowed?
+            }
+            else
+            {
+                if (outputFields.length() == 0)
+                    cmd->addId(wuTree->queryName());
+                else
+                {
+                    StringBuffer output = wuTree->queryName();
+                    //Append more into the output.
+                    setWUDataTree(wuTree, outputFields, output);
+                    cmd->addId(output.str());
                 }
             }
+            return true;
         }
-    }
-    if (cmd->getOnline()) {
-        if (haswusoutput)
-            throw MakeStringException(-1,"SCA_WORKUNIT_SERVICES_GET not implemented for online workunits!");
-        StringBuffer xpath(baseXPath);
-        if (!conn)
+        void setWUDataTree(IPropertyTree *wu, StringArray &outputFields, StringBuffer &output)
         {
-            if (!isWild)
-                xpath.append("/").append(mask);
-            conn.setown(querySDS().connect(xpath.str(), myProcessSession(), 0, 5*60*1000));
+            ForEachItemIn(i, outputFields)
+            {
+                const char* outputField = (const char*)outputFields.item(i);
+
+                StringBuffer val;
+                bool found = true;
+                if (strieq(outputField, "owner"))
+                    wu->getProp("@submitID", val);
+                else if (strieq(outputField, "cluster"))
+                    wu->getProp("@clusterName", val);
+                else if (strieq(outputField, "jobname"))
+                    wu->getProp("@jobName",val);
+                else if (strieq(outputField,"state"))
+                    wu->getProp("@state", val);
+                else
+                    found = false;
+                if (found)
+                    output.append(',').append(val);
+            }
         }
-        if (conn)
+        void addWUsSorted(StringArray &outputFields, IArrayOf<IPropertyTree> &wus)
         {
-            Owned<IPropertyTreeIterator> iter = conn->queryRoot()->getElements(isWild ? "*" : NULL);
-            unsigned index = 0;
-            StringBuffer val;
-            ForEach(*iter) {
-                IPropertyTree &pt=iter->query();
-                const char *wuid = pt.queryName();
-                if (index>start+num)
+            unsigned nwus = wus.ordinality();
+            UnsignedArray idx;
+            unsigned i;
+            for (i = 0; i < nwus; i++)
+                idx.append(i);
+            {
+                CriticalBlock block(wuDataSortSect);
+                wusRef = &wus;
+                wuDataSortInc = sortInc;
+                idx.sort(compare);
```

Comment:
As I say (here: https://github.com/hpcc-systems/HPCC-Platform/pull/10711#discussion_r162132190), globals and locks should not be necessary..

You don't need to build up a temporary array, set globals and use critical sections.
You can define a sort function and call wus.sort(suitableSortFunc) on it directly..

There are a few example that do same, in fact one in this file, see compareBranch, it's used by a array.bAdd() call instead of an array.sort() call, but the array and prototype for the CompareFunc are same for bAdd or sort.

If you look in jptree.cpp there's a IPropertyTree example, search fror comparePropTrees

Reply:
Are you sure the 'sortInc' ordering difference is correct here?
Surely you always want it in ascending order.. ?

The getArchivedWUs() branch uses sortInc to ensure that the array coming out of the reversed directory iterators is put back in ascending (old to new) order..
As you have it here currently, if sortInc is true, the sort function would sort it from new->old and return it like that..

Reply:
> You can define a sort function and call wus.sort(suitableSortFunc) on it directly..
The wus is a tree array. The idx is an integer array. Will the wus.sort() be the same efficient as the idx.sort()?

Reply:
Yes, it will be more efficient (especially if concurrent request blocking on crit).
It's only sorting pointers, the current code will be performing lots of indirect calls to wus.item as well.

Reply:
"Surely you always want it in ascending order.. ?"
The ideal order should be descending (from new to old). People may be more interested in newer WUs than older WUs. If so, for online WUs, I have to use the iptiter_sort first. Then, reverse and filter ... @jakesmith any suggestion?

Reply:
The line:
```
sortInc = !beforeWU.isEmpty() && afterWU.isEmpty();
```
.. could do with a comment explaining the purpose and reason "sortInc" is set in this case.

Reply:
>The ideal order should be descending (from new to old). People may be more interested in newer WUs than older WUs. If so, for online WUs, I have to use the iptiter_sort first. Then, reverse and filter ... @jakesmith any suggestion?

My main problem with the code as it is, is afaics it's inconsistent, it looks like:
```
Archived when sortInc=true, final order = old->new
Online when sortInc=true,  final order = old->new
Archived when sortInc=false, final order = old->new
Online when sortInc=false,  final order = *new->old*
```

i.e. it looks like the return order from archived is always old->new , is that not true?
And for online it looks like it is dependent on sortInc.

What order are they supposed to be returned in, in each case?
For the online case, if you want old->new order returned, using iptiter_sort would avoid the need for you to manually sort, if you need new->old order returned, then you'll have to sort yourself.
I don't think you need to change when the filtering is done.
@wangkx

Reply:
The inconsistent sorting will be fixed. The sortInc may be renamed and will be commented.

Reply:
I may add a client parameter (Descending) to control the order of returning WUs. The default order is new->old (Descending = true). The sortInc may be changed to countBackward. It is used to support paging. For example, a client retrieves archived WUs with the order new->old (Descending = true) and the client is on page 50. If a client wants to get the next page by specifying the afterWU, we may call the getSortedDirectoryIterator() with the order new->old, find the WU after the 'afterWU' and count a page of WUs. If a client wants to go back to the previous page by specifying the beforeWU, I may set countBackward = true (old->new), call the getSortedDirectoryIterator() with the order old->new, find the WU before the 'beforeWU', count a page of WUs, and reorder them to new->old. For onlineWUs, the logic may be similar but using iptiter_sort (for old->new) and manual sort (for new->old). It may take time to find the 'afterWU' or 'beforeWU' by iterating the sorted WUs. Similar logic will be implemented for Descending = false. @jakesmith any comment?

Reply:
@wangkx - I think that sounds fine.